examples/fp_delegation.cpp

```diff
@@ -0,0 +1,177 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unifex/for_each.hpp>
+#include <unifex/manual_event_loop.hpp>
+#include <unifex/range_stream.hpp>
+#include <unifex/scheduler_concepts.hpp>
+#include <unifex/sync_wait.hpp>
+#include <unifex/timed_single_thread_context.hpp>
+#include <unifex/transform_stream.hpp>
+#include <unifex/via_stream.hpp>
+#include <unifex/with_query_value.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <variant>
+
+using namespace unifex;
+using namespace std::chrono_literals;
+
+namespace {
+
+class delegating_scheduler;
+
+class delegating_context {
+  public:
+  delegating_context(int capacity) : capacity_{capacity} {}
+
+  bool reserve() {
+    auto lck = std::scoped_lock{m_};
+    bool reservation_made = reservation_count_ < capacity_;
+    if(reservation_made) {
+      ++reservation_count_;
+    }
+    return reservation_made;
+  }
+
+  int get_count() {
+    return run_count_;
+  }
+
+  void run() {
+    run_count_++;
+  }
+
+  delegating_scheduler get_scheduler() noexcept;
+
+  private:
+  friend class delegating_sender;
+  std::mutex m_;
+  int reservation_count_{0};
+  std::atomic<int> run_count_{0};
+  int capacity_;
+  timed_single_thread_context single_thread_context_;
+};
+
+template <typename DelegatedOperationState, typename LocalOperationState>
+class delegating_operation final {
+  public:
+  template<class InitFunc>
+  delegating_operation(InitFunc&& func, delegating_context* context) :
+    op_{std::in_place_type_t<remove_cvref_t<decltype(func())>>{}, func()}, context_{context} {
+  }
+
+
+  inline void start() noexcept {
+    if(std::holds_alternative<DelegatedOperationState>(op_)) {
+      // Start a delegated operation
+      std::get<DelegatedOperationState>(op_).start();
+    } else {
+      // Start immediately on the local context
+      context_->run();
+      std::get<LocalOperationState>(op_).start();;
+    }
+  }
+
+  std::variant<DelegatedOperationState, LocalOperationState> op_;
+  delegating_context* context_ = nullptr;
+};
+
+class delegating_sender {
+  public:
+  template <
+      template <typename...> class Variant,
+      template <typename...> class Tuple>
+  using value_types = Variant<Tuple<>>;
+
+  template <template <typename...> class Variant>
+  using error_types = Variant<>;
+
+  template<typename OpType>
+  struct LocalContextType {
+    OpType op_;
+
+    inline void start() noexcept {
+      op_.start();
+    }
+  };
+
+  template <typename Receiver>
+  auto connect(Receiver&& receiver) {
```

Comment:
Would you also be able to implement this pattern with, say, a `fallback_scheduler()`?

eg. if we have a capacity-constrained scheduler like you have here, which has a schedule() operation that might fail with an error, we could turn it into a delegating scheduler by building a `fallback_scheduler(capacity_constrainted_scheduler(), current_scheduler())` where `current_scheduler` is a placeholder that forwards `schedule()` to the scheduler returned from the receiver's `get_scheduler()`.

I don't think it's quite as general as a scheduler that explicitly uses the calling context's scheduler, but it would seem to support the use-case in this example.

Reply:
I think that would work. Are you asking for an implementation or just thinking aloud?

Reply:
Mainly just thinking aloud.

I think for this to address the forward-progress requirements more generally, though, if we want to allow the user to pass an arbitrary executor to use for a parallel algorithm, then we'll need to create a different kind of 

This one assumes that we only need to fallback to the sync_wait() thread's executor if the other executor's schedule() completes with set_error(). However, if the other executor has accepted the work but just not scheduled it yet (eg. all its threads are currently blocked on other things) then we'd still want the sync_wait() executor to be able to pick up the work and execute it and to be able to cancel the work scheduled to the user-provided executor if the sync_wait() executor finishes the work first.